File: wearos/app/src/main/java/com/precpet/wearos/session/SessionModels.kt
package com.precpet.wearos.session

/**
 * Phone-free offline session models.
 *
 * A recorded session stores the exact Precept Motion Service packets (the same
 * 16-byte frames the BLE streamer emits, see `docs/wearable-protocol.md` §4) as
 * a base64 blob, so the PWA can decode every sample with the same
 * `decodeImuPacket` used for live streaming and feed it through the shared
 * `analyzeMotion` pipeline.
 */
data class SessionSummary(
    val id: String,
    val startedAtMs: Long,
    val endedAtMs: Long,
    val sampleCount: Int,
    val avgAccelMagnitude: Double, // m/s² (RMS, includes gravity)
    val peakGyroMagnitude: Float, // deg/s
    val repCount: Int = 0, // on-watch rep detector (§12); 0 = not counted
) {
    val durationMs: Long get() = endedAtMs - startedAtMs
}

data class StoredSession(
    val summary: SessionSummary,
    val packetVersion: Int,
    val packetSize: Int,
    val samplesBase64: String,
) {
    val sampleRate: Double
        get() = if (summary.durationMs > 0) {
            summary.sampleCount * 1000.0 / summary.durationMs
        } else {
            0.0
        }
}

File: watchos/PreceptWatchOS/Session/SessionModels.swift
import Foundation

/// Phone-free offline session models.
/// Mirrors the Wear OS `SessionModels.kt` and the JSON shape consumed by the
/// PWA over BLE (docs/wearable-protocol.md §12).
struct SessionSummary: Codable {
    let id: String
    let startedAtMs: Int64
    let endedAtMs: Int64
    let sampleCount: Int
    let avgAccelMagnitude: Double // m/s² (RMS, includes gravity)
    let peakGyroMagnitude: Float // deg/s
    var repCount: Int = 0 // on-watch rep detector (§12); 0 = not counted

    var durationMs: Int64 { endedAtMs - startedAtMs }
}

struct StoredSession: Codable {
    let v: Int
    let id: String
    let startedAtMs: Int64
    let endedAtMs: Int64
    let sampleCount: Int
    let avgAccelMagnitude: Double
    let peakGyroMagnitude: Float
    let repCount: Int
    let packetVersion: Int
    let packetSize: Int
    let samplesBase64: String

    init(summary: SessionSummary, packetVersion: Int, packetSize: Int, samplesBase64: String) {
        self.v = 1
        self.id = summary.id
        self.startedAtMs = summary.startedAtMs
        self.endedAtMs = summary.endedAtMs
        self.sampleCount = summary.sampleCount
        self.avgAccelMagnitude = summary.avgAccelMagnitude
        self.peakGyroMagnitude = summary.peakGyroMagnitude
        self.repCount = summary.repCount
        self.packetVersion = packetVersion
        self.packetSize = packetSize
        self.samplesBase64 = samplesBase64
    }

    var summary: SessionSummary {
        SessionSummary(
            id: id,
            startedAtMs: startedAtMs,
            endedAtMs: endedAtMs,
            sampleCount: sampleCount,
            avgAccelMagnitude: avgAccelMagnitude,
            peakGyroMagnitude: peakGyroMagnitude,
            repCount: repCount
        )
    }

    var sampleRate: Double {
        durationMs > 0 ? Double(sampleCount) * 1000.0 / Double(durationMs) : 0
    }
}
